src/catalog/database_catalog.cpp

```diff
@@ -30,856 +29,89 @@
-                            postgres::CONSTRAINT_INDEX_INDEX_OID, "pg_constraint_index_index",
-                            postgres::Builder::GetConstraintIndexIndexSchema(db_oid_));
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-  retval = SetIndexPointer(txn, postgres::CONSTRAINT_INDEX_INDEX_OID, constraints_index_index_);
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-
-  retval = CreateIndexEntry(txn, postgres::NAMESPACE_CATALOG_NAMESPACE_OID, postgres::CONSTRAINT_TABLE_OID,
-                            postgres::CONSTRAINT_FOREIGNTABLE_INDEX_OID, "pg_constraint_foreigntable_index",
-                            postgres::Builder::GetConstraintForeignTableIndexSchema(db_oid_));
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-  retval = SetIndexPointer(txn, postgres::CONSTRAINT_FOREIGNTABLE_INDEX_OID, constraints_foreigntable_index_);
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-
-  // pg_language and associated indexes
-  retval = CreateTableEntry(txn, postgres::LANGUAGE_TABLE_OID, postgres::NAMESPACE_CATALOG_NAMESPACE_OID, "pg_language",
-                            postgres::Builder::GetLanguageTableSchema());
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-  retval = SetTablePointer(txn, postgres::LANGUAGE_TABLE_OID, languages_);
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-
-  retval = CreateIndexEntry(txn, postgres::NAMESPACE_CATALOG_NAMESPACE_OID, postgres::LANGUAGE_TABLE_OID,
-                            postgres::LANGUAGE_OID_INDEX_OID, "pg_languages_oid_index",
-                            postgres::Builder::GetLanguageOidIndexSchema(db_oid_));
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-  retval = SetIndexPointer(txn, postgres::LANGUAGE_OID_INDEX_OID, languages_oid_index_);
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-
-  retval = CreateIndexEntry(txn, postgres::NAMESPACE_CATALOG_NAMESPACE_OID, postgres::LANGUAGE_TABLE_OID,
-                            postgres::LANGUAGE_NAME_INDEX_OID, "pg_languages_name_index",
-                            postgres::Builder::GetLanguageNameIndexSchema(db_oid_));
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
-  retval = SetIndexPointer(txn, postgres::LANGUAGE_NAME_INDEX_OID, languages_name_index_);
-  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
+void DatabaseCatalog::TearDown(const common::ManagedPointer<transaction::TransactionContext> txn) {
+  auto teardown_pg_core = pg_core_.GetTearDownFn(txn, common::ManagedPointer(this));
+  auto teardown_pg_constraint = pg_constraint_.GetTearDownFn(txn);
+  auto teardown_pg_proc = pg_proc_.GetTearDownFn(txn);
 
-  BootstrapLanguages(txn);
+  auto dbc_nuke = [=]() {
+    // Order matters!
+    teardown_pg_core();
+    teardown_pg_constraint();
+    teardown_pg_proc();
```

Comment:
If order matters, then I'm surprised to see them in this order rather than reverse bootstrap order.  I'm pretty sure order does not matter here since the deferrals should ensure any live references across the catalog are dead before the object deletions occur.  If I'm wrong, this needs a more explicit comment on the ordering.

Reply:
You are right, it teardowns fine if I flip ordering around. I don't remember why I wrote that (whoops).

Reply:
removed